lib/logstash/outputs/dynatrace.rb

```diff
@@ -73,12 +73,39 @@ def headers
       def multi_receive(events)
         return if events.length.zero?
 
-        request = Net::HTTP::Post.new(uri, headers)
-        request.body = "#{LogStash::Json.dump(events.map(&:to_hash)).chomp}\n"
-        response = @client.request(request)
-        return if response.is_a? Net::HTTPSuccess
+        retries = 0
+        begin
+          request = Net::HTTP::Post.new(uri, headers)
+          request.body = "#{LogStash::Json.dump(events.map(&:to_hash)).chomp}\n"
+          response = send(request)
+          return if response.is_a? Net::HTTPSuccess
+
+          failure_message = "Dynatrace returned #{response.code} #{response.message}."
+
+          if response.is_a? Net::HTTPServerError
+            raise RetryableError.new failure_message
+          end
+
+          if response.is_a? Net::HTTPClientError
+            @logger.error("#{failure_message} Please check that log ingest is enabled and your API token has the `logs.ingest` (Ingest Logs) scope.")
+            return
+          end
+        rescue Net::HTTPBadResponse, RetryableError => e
+          # indicates a protocol error
+          if retries < MAX_RETRIES  
+            sleep_seconds = 2 ** retries
+            @logger.warn("Failed to contact dynatrace: #{e.message}. Trying again after #{sleep_seconds} seconds.")
+            sleep sleep_seconds
```

Comment:
Is it ensured that the exporter is always running on a dedicated thread?

Reply:
I'm fairly sure yes but i'll check

Reply:
There is a pool of worker threads. This will block one of the threads. The number of threads is configurable but is set to the number of CPU cores by default.

Reply:
> When configured as :single a single instance of the Output will be shared among the
> pipeline worker threads. Access to the `#multi_receive/#multi_receive_encoded/#receive` method will be synchronized
> i.e. only one thread will be active at a time making threadsafety much simpler.

This means that we will only block at most 1 thread. We still have to worry about exhausting the queue, but by default the queue is an unlimited in-memory queue.